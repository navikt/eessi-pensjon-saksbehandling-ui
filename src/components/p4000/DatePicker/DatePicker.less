div.toggle {
    display: inline-flex !important;
}

.react-date-picker__button {
    border-radius: 5px;
    background-color: white;
}

.react-date-picker__button__input {
    padding: 5px;
}

.react-date-picker--disabled {
    background-color: darkgrey;
}

.toggle__label {
    height: 100% !important;
    max-width: 10rem;
}
