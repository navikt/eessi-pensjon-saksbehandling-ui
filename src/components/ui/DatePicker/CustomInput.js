import React, { Component } from 'react'
import PT from 'prop-types'
import classNames from 'classnames'

class CustomInput extends Component {
  render () {
    const { onClick, onChange, value, placeholder, error, id } = this.props
    return <div className='input-group'>
      <input
        id={id}
        type='text'
        className={classNames('form-control', { 'skjemaelement__input--harFeil': error })}
        placeholder={placeholder}
        onChange={onChange}
        value={value} />
      <div style={{ cursor: 'pointer' }} className='input-group-append' onClick={onClick}>
        <span role='img' aria-label='date' className='input-group-text'>ðŸ“…</span>
      </div>
    </div>
  }
}

CustomInput.propTypes = {
  onClick: PT.func,
  value: PT.string
}

export default CustomInput
