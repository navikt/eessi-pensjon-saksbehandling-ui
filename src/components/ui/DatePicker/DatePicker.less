@import '../../../constants';

.react-datepicker-popper {
  z-index: 2 !important:
}

.highContrast .c-ui-datePicker {
  input {
    background: @black;
  }
  .input-group-append span {
    background: @black;
  }
}
.highContrast .react-datepicker-popper {
  .react-datepicker,
  .react-datepicker__header,
  .react-datepicker__month-select,
  .react-datepicker__year-select {
    background-color: @black;
  }
}
