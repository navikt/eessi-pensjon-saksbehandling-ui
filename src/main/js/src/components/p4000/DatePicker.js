import React, { Component } from 'react';
import PT from 'prop-types';
import { translate } from 'react-i18next';

import YearMonthSelector from 'react-year-month-selector';
import 'react-year-month-selector/src/styles/index.css';
import './custom-datepicker.css';
import * as Nav from '../ui/Nav';

class DatePicker extends Component {

    state = {
        rangePeriod: true
    };

    onStartDateChange(year, month) {

        let { t, onStartDatePicked } = this.props;
        let fixedMonth = month + 1;
        this.setState({
            startYear : year,
            startMonth: fixedMonth,
            onStartDate: year + '/' + fixedMonth
        }, () => {
            if (this.state.endYear < this.state.startYear ||
             (this.state.endYear === this.startYear && this.state.endMonth < this.state.startMonth)) {
                 this.setState({
                      onEndDateFail: t('ui:onEndDateFail')
                 })
             } else {
                  this.setState({
                      onEndDateFail: undefined
                  })
             }
             onStartDatePicked(year, fixedMonth);
        });
    }

    onEndDateChange(year, month) {

        let { t, onEndDatePicked } = this.props;
        let fixedMonth = month + 1;
        this.setState({
            endYear : year,
            endMonth: fixedMonth,
            onEndDate: year + '/' + fixedMonth
        }, () => {
           if (this.state.endYear < this.state.startYear ||
            (this.state.endYear === this.startYear && this.state.endMonth < this.state.startMonth)) {
                this.setState({
                     onEndDateFail: t('validation:endDateEarlierThanStartDate')
                })
            } else {
                 this.setState({
                     onEndDateFail: undefined
                 })
            }
            onEndDatePicked(year, fixedMonth);
        });
    }

    onFocus (input) {

        this.setState({[input]: true});
    }

    onClose(input) {

        this.setState({[input]: false});
    }

    setRangePeriod() {
        this.setState({
            rangePeriod: true
        })
    }

    setOpenPeriod() {
        this.setState({
            rangePeriod: false
        })
    }

    render () {

        let { t, onStartDatePicked, onEndDatePicked } = this.props;

        return <div>
        <Nav.Row>
            <Nav.Column>
                <Nav.Radio label={t('ui:rangePeriod')} name='period' checked={this.state.rangePeriod === true}
                    onChange={this.setRangePeriod.bind(this)}/>
                <Nav.Radio label={t('ui:openPeriod')} name='period' checked={this.state.rangePeriod === false}
                    onChange={this.setOpenPeriod.bind(this)}/>
            </Nav.Column>
        </Nav.Row>
        <Nav.Row>
            <Nav.Column>
                <Nav.Input label={t('ui:startDate')} value={this.state.onStartDate}
                onFocus={this.onFocus.bind(this, 'startopen')}
                feil={this.state.onStartDateFail ? {'feilmelding' : this.state.onStartDateFail} : null}/>
                <YearMonthSelector onChange={this.onStartDateChange.bind(this)} open={this.state.startopen}
                onClose={this.onClose.bind(this, 'startopen')}/>
            </Nav.Column>
            {this.state.rangePeriod ? <Nav.Column>
                    <Nav.Input label={t('ui:endDate')} value={this.state.onEndDate}
                    onFocus={this.onFocus.bind(this, 'endopen')}
                    feil={this.state.onEndDateFail ? {'feilmelding' : this.state.onEndDateFail} : null}/>
                    <YearMonthSelector onChange={this.onEndDateChange.bind(this)} open={this.state.endopen}
                    onClose={this.onClose.bind(this, 'endopen')}/>
                </Nav.Column> : null}
        </Nav.Row>
        </div>
    }
}

DatePicker.propTypes = {
    onStartDatePicked : PT.func.isRequired,
    onEndDatePicked   : PT.func.isRequired,
    t                 : PT.func.isRequired
};

export default translate()(DatePicker);

