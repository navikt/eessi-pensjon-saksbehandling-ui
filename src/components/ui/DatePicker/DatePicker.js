import React, { Component } from 'react'
import PT from 'prop-types'
import classNames from 'classnames'

import CustomInput from './CustomInput'

import ReactDatePicker from 'react-datepicker'
import 'react-datepicker/dist/react-datepicker.min.css'

class DatePicker extends Component {
  render () {
    const { selected, className, locale, error, onChange, placeholder } = this.props

    return <React.Fragment>
      <ReactDatePicker
        customInput={<CustomInput {...this.props} />}
        selected={selected}
        className={classNames('input-group', className)}
        dateFormat='dd.MM.yyyy'
        showYearDropdown
        showMonthDropdown
        dropdownMode='select'
        locale={locale}
        onChange={onChange} />
      {error ? <div>{error}</div> : null}
    </React.Fragment>
  }
}

DatePicker.propTypes = {

}

export default DatePicker
