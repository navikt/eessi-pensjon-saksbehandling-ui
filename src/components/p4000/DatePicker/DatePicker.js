import React, { Component } from 'react';
import PT from 'prop-types';
import { translate } from 'react-i18next';
import { connect } from 'react-redux';
import { bindActionCreators }  from 'redux';
import classNames from 'classnames';
import ReactDatePicker from 'react-datepicker';
import 'react-datepicker/dist/react-datepicker.min.css';
import moment from 'moment';

import Validation from '../Validation';
import * as Nav from '../../ui/Nav';
import * as p4000Actions from '../../../actions/p4000';

import './DatePicker.css';

const mapStateToProps = (state) => {
    return {
        event  : state.p4000.event,
        locale : state.ui.locale
    }
};

const mapDispatchToProps = (dispatch) => {
    return {actions: bindActionCreators(Object.assign({}, p4000Actions), dispatch)};
};

class DatePicker extends Component {

    state = {
        validationError: undefined
    };

    componentDidMount() {

        const { event, actions } = this.props;

        if (!event.dateType) {
            // set initial state
            actions.setEventProperty('dateType', 'both');
        }
        this.props.provideController({
            hasNoValidationErrors : this.hasNoValidationErrors.bind(this),
            passesValidation      : this.passesValidation.bind(this),
            resetValidation       : this.resetValidation.bind(this)
        });
    }

    componentWillUnmount() {

        this.props.provideController(null)
    }

    hasNoValidationErrors() {

        return this.state.validationError === undefined;
    }

    async resetValidation() {

        return new Promise(async (resolve, reject) => {

            try {
                this.setState({
                    validationError: undefined
                }, () => {
                    // after setting up state, use it to see the validation state
                    resolve();
                });
            } catch (error) {
                reject(error);
            }
        });
    }

    async passesValidation() {

        const { event } = this.props;

        return new Promise((resolve, reject) => {

            try {
                this.setState({
                    validationError : Validation.validateDatePicker(event.dateType, event)
                }, () => {
                    // after setting up state, use it to see the validation state
                    resolve(this.hasNoValidationErrors());
                });
            } catch (error) {
                reject(error);
            }
        });
    }

    handlePeriodChange(e) {

        let { event, actions } = this.props;
        let dateType = e.target.value;

        actions.setEventProperty('dateType', dateType);
        if (dateType !== 'both' && event.endDate) {
            actions.setEventProperty('endDate', undefined);
        }
    }

    onStartDateBlur(e) {

        const { event, actions } = this.props;
        let date = e.target.value;

        if (! /\d\d\.\d\d\.\d\d\d\d/.test(date)) {

            if (!event.startDate || date !== event.startDate)  {

                this.setState({
                    validationError : 'p4000:validation-invalidDate'
                },  () => {
                    actions.setEventProperty('startDate', undefined);
                });
            }
        } else {
            let startDate = moment(date, 'DD.MM.YYYY').toDate();
            if (!event.startDate || startDate.getTime() !== event.startDate.getTime())  {
                this.onStartDateHandle(startDate);
            }
        }
    }

    onEndDateBlur(e) {

        const { event, actions } = this.props;
        let date = e.target.value;

        if (! /\d\d\.\d\d\.\d\d\d\d/.test(date)) {

            if (!event.endDate || date !== event.endDate)  {

                this.setState({
                    validationError : 'p4000:validation-invalidDate'
                },  () => {
                    actions.setEventProperty('endDate', undefined);
                });
            }
        } else {
            let endDate = moment(date, 'DD.MM.YYYY').toDate();
            if (!event.endDate || endDate.getTime() !== event.endDate.getTime())  {
                this.onEndDateHandle(endDate);
            }
        }
    }

    onStartDateChange(moment) {

        const { event } = this.props;

        let date = moment.toDate();
        if (!event.startDate || date.getTime() !== event.startDate.getTime())  {
            this.onStartDateHandle(date);
        }
    }

    onEndDateChange(moment) {

        const { event } = this.props;

        let date = moment.toDate();
        if (!event.endDate || date.getTime() !== event.endDate.getTime())  {
            this.onEndDateHandle(date);
        }
    }

    onStartDateHandle(date) {

        let { event, actions } = this.props;
        let validationError = undefined;

        if (event.dateType === 'both') {
            if (event.endDate && event.endDate < date) {
                validationError = 'p4000:validation-endDateEarlierThanStartDate';
            }
        }
        if (date > new Date()) {
            validationError = 'p4000:validation-startDateCantBeInFuture';
        }

        this.setState({
            validationError : validationError
        }, () => {
            actions.setEventProperty('startDate', date);
        });
    }

    onEndDateHandle(date) {

        let { event, actions } = this.props;
        let validationError = undefined;

        if (event.dateType === 'both') {
            if (event.startDate && event.startDate > date) {
                validationError = 'p4000:validation-endDateEarlierThanStartDate';
            }
        }
        if (date > new Date()) {
            validationError = 'p4000:validation-endDateCantBeInFuture';
        }

        this.setState({
            validationError : validationError
        }, () => {
            actions.setEventProperty('endDate', date);
        });
    }

    handleUncertainDateChange(e) {

        let { actions } = this.props;
        actions.setEventProperty('uncertainDate', e.target.checked);
    }

    getToggle(dateType) {

        let { t } = this.props;

        return <Nav.ToggleGruppe name='dateType' className='dateType' style={{display: 'inline-flex'}} onChange={this.handlePeriodChange.bind(this)}>
            <Nav.ToggleKnapp value='both'            checked={dateType === 'both'}            key='1'>{t('p4000:form-rangePeriod')}</Nav.ToggleKnapp>
            <Nav.ToggleKnapp value='onlyStartDate01' checked={dateType === 'onlyStartDate01'} key='2'>{t('p4000:form-onlyStartDate01')}</Nav.ToggleKnapp>
            <Nav.ToggleKnapp value='onlyStartDate98' checked={dateType === 'onlyStartDate98'} key='3'>{t('p4000:form-onlyStartDate98')}</Nav.ToggleKnapp>
        </Nav.ToggleGruppe>
    }

    render () {

        let { t, event, locale } = this.props;

        return <div className={classNames('div-datePicker', 'p-2')}>
            {!this.hasNoValidationErrors() ? <Nav.AlertStripe className='mb-3' type='advarsel'>{t(this.state.validationError)}</Nav.AlertStripe> : null}
            <Nav.Row className='row-datePicker-toggleButton no-gutters'>
                <Nav.Column className='text-center mb-4'>
                    {this.getToggle(event.dateType)}
                    <Nav.Checkbox className='d-inline-flex ml-4 mt-3 uncertainDate'
                        label={t('p4000:form-uncertainDate')}
                        checked={event.uncertainDate}
                        onChange={this.handleUncertainDateChange.bind(this)}/>
                </Nav.Column>
            </Nav.Row>
            <Nav.Row className='row-datepickers no-gutters'>
                <Nav.Column className='text-center'>
                    <label className='mr-3'>{t('ui:startDate') + ' *'}</label>
                    <ReactDatePicker selected={event.startDate ? moment(event.startDate) : undefined}
                        className='startDate'
                        dateFormat='DD.MM.YYYY'
                        placeholderText={t('ui:dateFormat')}
                        showYearDropdown
                        showMonthDropdown
                        dropdownMode='select'
                        locale={locale}
                        onMonthChange={this.onStartDateChange.bind(this)}
                        onYearChange={this.onStartDateChange.bind(this)}
                        onBlur={this.onStartDateBlur.bind(this)}
                        onChange={this.onStartDateChange.bind(this)}/>
                    <div>{this.state.onStartDateFail}</div>
                </Nav.Column>
                <Nav.Column className='text-center'>
                    <label className='mr-3'>{t('ui:endDate')}</label>
                    <ReactDatePicker selected={event.endDate ? moment(event.endDate) : undefined}
                        disabled={event.dateType !== 'both'}
                        className='endDate'
                        dateFormat='DD.MM.YYYY'
                        placeholderText={t('ui:dateFormat')}
                        showYearDropdown
                        showMonthDropdown
                        dropdownMode='select'
                        locale={locale}
                        onMonthChange={this.onEndDateChange.bind(this)}
                        onYearChange={this.onEndDateChange.bind(this)}
                        onBlur={this.onEndDateBlur.bind(this)}
                        onChange={this.onEndDateChange.bind(this)}/>
                    <div>{this.state.onEndDateFail}</div>
                </Nav.Column>
            </Nav.Row>
        </div>
    }
}

DatePicker.propTypes = {
    t                 : PT.func.isRequired,
    event             : PT.object.isRequired,
    actions           : PT.object.isRequired,
    provideController : PT.func.isRequired,
    locale            : PT.string.isRequired
};

export default connect(
    mapStateToProps,
    mapDispatchToProps
)(
    translate()(DatePicker)
);
