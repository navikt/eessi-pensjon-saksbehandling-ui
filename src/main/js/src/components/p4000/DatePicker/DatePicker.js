import React, { Component } from 'react';
import PT from 'prop-types';
import { translate } from 'react-i18next';
import { connect } from 'react-redux';
import { bindActionCreators }  from 'redux';
import classNames from 'classnames';
import ReactDatePicker from 'react-date-picker';

import Validation from '../Validation';
import * as Nav from '../../ui/Nav';
import * as p4000Actions from '../../../actions/p4000';

import './DatePicker.css';

const mapStateToProps = (state) => {
    return {
        event : state.p4000.event
    }
};

const mapDispatchToProps = (dispatch) => {
    return {actions: bindActionCreators(Object.assign({}, p4000Actions), dispatch)};
};

class DatePicker extends Component {

    state = {
        dateType: 'both',
        validationError: undefined
    };

    componentDidMount() {

        this.props.provideController({
            hasNoValidationErrors : this.hasNoValidationErrors.bind(this),
            passesValidation      : this.passesValidation.bind(this),
            resetValidation       : this.resetValidation.bind(this)
        });
    }

    componentWillUnmount() {

        this.props.provideController(null)
    }

    hasNoValidationErrors() {

        return this.state.validationError === undefined;
    }

    async resetValidation() {

        return new Promise(async (resolve, reject) => {

            try {
                this.setState({
                    validationError: undefined
                }, () => {
                    // after setting up state, use it to see the validation state
                    resolve();
                });
            } catch (error) {
                reject(error);
            }
        });
    }

    async passesValidation() {

        const { event } = this.props;

        return new Promise((resolve, reject) => {

            try {
                this.setState({
                    validationError : Validation.validateDatePicker(this.state.dateType, event)
                }, () => {
                    // after setting up state, use it to see the validation state
                    resolve(this.hasNoValidationErrors());
                });
            } catch (error) {
                reject(error);
            }
        });
    }

    handlePeriodChange(e) {

        let { event, actions } = this.props;
        let dateType = e.target.value;

        this.setState({
            dateType: dateType
        }, () => {
            actions.setEventProperty('dateType', dateType);
            if (dateType !== 'both' && event.endDate) {
                actions.setEventProperty('endDate', undefined);
            }
        });
    }

    onStartDateChange(date) {

        let { event, actions } = this.props;
        let validationError = undefined;

        if (this.state.dateType === 'both') {
            if (event.endDate && event.endDate < date) {
                validationError = 'p4000:validation-endDateEarlierThanStartDate';
            }
        }
        if (date > new Date()) {
            validationError = 'p4000:validation-startDateCantBeInFuture';
        }

        this.setState({
            validationError : validationError
        }, () => {
            actions.setEventProperty('dateType', this.state.dateType);
            actions.setEventProperty('startDate', date);
        });
    }

    onEndDateChange(date) {

        let { event, actions } = this.props;
        let validationError = undefined;

        if (this.state.dateType === 'both') {
            if (event.startDate && event.startDate > date) {
                validationError = 'p4000:validation-endDateEarlierThanStartDate';
            }
        }
        if (date > new Date()) {
            validationError = 'p4000:validation-endDateCantBeInFuture';
        }

        this.setState({
            validationError : validationError
        }, () => {
            actions.setEventProperty('dateType', this.state.dateType);
            actions.setEventProperty('endDate', date);
        });
    }

    handleUncertainDateChange(e) {

        let { actions } = this.props;
        actions.setEventProperty('uncertainDate', e.target.value);
    }

    render () {

        let { t, event } = this.props;

        return <div className={classNames('div-datePicker', 'p-2')}>
            {!this.hasNoValidationErrors() ? <Nav.AlertStripe className='mb-3' type='advarsel'>{t(this.state.validationError)}</Nav.AlertStripe> : null}
            <Nav.Row className='row-datePicker-toggleButton no-gutters'>
                <Nav.Column className='text-center mb-4'>
                    <Nav.ToggleGruppe name='datePickerType' style={{display: 'inline-flex'}} onChange={this.handlePeriodChange.bind(this)}>
                        <Nav.ToggleKnapp value='both' defaultChecked={true} key='1'>{t('p4000:rangePeriod')}</Nav.ToggleKnapp>
                        <Nav.ToggleKnapp value='onlyStartDate01' defaultChecked={false} key='2'>{t('p4000:onlyStartDate01')}</Nav.ToggleKnapp>
                        <Nav.ToggleKnapp value='onlyStartDate98' defaultChecked={false} key='3'>{t('p4000:onlyStartDate98')}</Nav.ToggleKnapp>
                    </Nav.ToggleGruppe>
                    <Nav.Checkbox className='d-inline-flex ml-4 mt-3' label={t('p4000:uncertainDate')}
                        checked={event.uncertainDate}
                        onChange={this.handleUncertainDateChange.bind(this)}/>
                </Nav.Column>
            </Nav.Row>
            <Nav.Row className='row-datepickers no-gutters'>
                <Nav.Column className='text-center'>
                    <label className='mr-3'>{t('ui:startDate') + ' *'}</label>
                    <ReactDatePicker value={event.startDate}
                        locale='no-NB'
                        onChange={this.onStartDateChange.bind(this)}/>
                    <div>{this.state.onStartDateFail}</div>
                </Nav.Column>
                <Nav.Column className='text-center'>
                    <label className='mr-3'>{t('ui:endDate')}</label>
                    <ReactDatePicker value={event.endDate} disabled={this.state.dateType !== 'both'}
                        locale='no-NB'
                        onChange={this.onEndDateChange.bind(this)}/>
                    <div>{this.state.onEndDateFail}</div>
                </Nav.Column>
            </Nav.Row>
        </div>
    }
}

DatePicker.propTypes = {
    t                 : PT.func.isRequired,
    event             : PT.object.isRequired,
    actions           : PT.object.isRequired,
    provideController : PT.func.isRequired
};

export default connect(
    mapStateToProps,
    mapDispatchToProps
)(
    translate()(DatePicker)
);
