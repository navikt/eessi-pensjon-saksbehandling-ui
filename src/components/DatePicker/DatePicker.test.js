import React from 'react'
import DatePicker from './DatePicker'

describe('DatePicker', () => {
  it('renders without crashing', () => {
    const wrapper = shallow(<DatePicker onChange={function () { }} />)
    expect(wrapper).toMatchSnapshot()
  })

  it('Correctly validates day', () => {
    const wrapper = shallow(<DatePicker onChange={function () { }} />)
    expect(wrapper.instance().checkValidity({ day: 'DD' }).day).toBeTruthy()
    expect(wrapper.instance().checkValidity({ day: '-1' }).day).toBeTruthy()
    expect(wrapper.instance().checkValidity({ day: '0' }).day).toBeTruthy()
    expect(wrapper.instance().checkValidity({ day: '10' }).day).toEqual(undefined)
    expect(wrapper.instance().checkValidity({ day: 10 }).day).toEqual(undefined)
    expect(wrapper.instance().checkValidity({ day: '50' }).day).toBeTruthy()
    expect(wrapper.instance().checkValidity({ day: '31', month: '3' }).day).toEqual(undefined)
    expect(wrapper.instance().checkValidity({ day: '31', month: '4' }).day).toBeTruthy()
    expect(wrapper.instance().checkValidity({ day: '29', month: '2', year: '2001' }).day).toBeTruthy()
    expect(wrapper.instance().checkValidity({ day: '29', month: '2', year: '2000' }).day).toEqual(undefined)
    expect(wrapper.instance().checkValidity({ day: '28', month: '2', year: '2001' }).day).toEqual(undefined)
  })

  it('Correctly validates month', () => {
    const wrapper = shallow(<DatePicker onChange={function () {}} />)
    expect(wrapper.instance().checkValidity({ month: 'MM' }).month).toBeTruthy()
    expect(wrapper.instance().checkValidity({ month: '-1' }).month).toBeTruthy()
    expect(wrapper.instance().checkValidity({ month: '0' }).month).toBeTruthy()
    expect(wrapper.instance().checkValidity({ month: '1' }).month).toEqual(undefined)
    expect(wrapper.instance().checkValidity({ month: 1 }).month).toEqual(undefined)
    expect(wrapper.instance().checkValidity({ month: '12' }).month).toEqual(undefined)
    expect(wrapper.instance().checkValidity({ month: 12 }).month).toEqual(undefined)
    expect(wrapper.instance().checkValidity({ month: '13' }).month).toBeTruthy()
  })

  it('Correctly validates year', () => {
    const wrapper = shallow(<DatePicker onChange={function () {}} />)
    expect(wrapper.instance().checkValidity({ year: 'YYYY' }).year).toBeTruthy()
    expect(wrapper.instance().checkValidity({ year: '0' }).year).toEqual(undefined)
    expect(wrapper.instance().checkValidity({ year: 0 }).year).toEqual(undefined)
    expect(wrapper.instance().checkValidity({ year: '2000' }).year).toEqual(undefined)
    expect(wrapper.instance().checkValidity({ year: 2000 }).year).toEqual(undefined)
    expect(wrapper.instance().checkValidity({ year: '3000' }).year).toEqual(undefined)
    expect(wrapper.instance().checkValidity({ year: 3000 }).year).toEqual(undefined)
  })

  it('Renders errors correctly', () => {
    const wrapper = shallow(<DatePicker onChange={function () {}} />)

    expect(wrapper.exists({ label: 'dag', feil: { feilmelding: '' } })).toBeFalsy()
    expect(wrapper.exists({ label: 'm책ned', feil: { feilmelding: '' } })).toBeFalsy()
    expect(wrapper.exists({ label: '책r', feil: { feilmelding: '' } })).toBeFalsy()
    wrapper.setState({ errors: { day: true, month: true, year: true } })
    expect(wrapper.exists({ label: 'dag', feil: { feilmelding: '' } })).toBeTruthy()
    expect(wrapper.exists({ label: 'm책ned', feil: { feilmelding: '' } })).toBeTruthy()
    expect(wrapper.exists({ label: '책r', feil: { feilmelding: '' } })).toBeTruthy()

    expect(wrapper.exists('.skjemaelement__feilmelding')).toBeFalsy()
    wrapper.setProps({ onChange: function () {}, feil: { feilmelding: 'ERROR' } })
    expect(wrapper.exists('.skjemaelement__feilmelding')).toBeTruthy()
    expect(wrapper.find('.skjemaelement__feilmelding').text()).toEqual('ERROR')
  })

  it('Sets state correctly', () => {
    const wrapper = shallow(<DatePicker onChange={function () {}} values={{ day: '1', month: '1', year: '1970' }} />)
    wrapper.instance().dateChange('day', { target: { value: 'NaN' } })
    expect(wrapper.state().errors.day).toBeTruthy()
    expect(wrapper.state().errors.month).toEqual(undefined)
    expect(wrapper.state().errors.year).toEqual(undefined)
    wrapper.instance().dateChange('month', { target: { value: 'NaN' } })
    expect(wrapper.state().errors.day).toEqual(undefined)
    expect(wrapper.state().errors.month).toBeTruthy()
    expect(wrapper.state().errors.year).toEqual(undefined)
    wrapper.instance().dateChange('year', { target: { value: 'NaN' } })
    expect(wrapper.state().errors.day).toEqual(undefined)
    expect(wrapper.state().errors.month).toEqual(undefined)
    expect(wrapper.state().errors.year).toBeTruthy()
  })

  it('Returns correct value', (done) => {
    const wrapper = shallow(<DatePicker onChange={onChange} values={{ day: '1', month: '1', year: '1970' }} />)

    function onChange (value) {
      expect(value).toEqual(objectToTestAgainst)
      done()
    }

    let objectToTestAgainst

    objectToTestAgainst = { day: '31', month: '1', year: '1970' }
    wrapper.instance().dateChange('day', { target: { value: '31' } })

    objectToTestAgainst = { day: '1', month: '12', year: '1970' }
    wrapper.instance().dateChange('month', { target: { value: '12' } })

    objectToTestAgainst = { day: '1', month: '1', year: '2010' }
    wrapper.instance().dateChange('year', { target: { value: '2010' } })
  })
})
