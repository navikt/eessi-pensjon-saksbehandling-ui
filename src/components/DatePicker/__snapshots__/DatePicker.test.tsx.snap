// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`components/DatePicker Renders 1`] = `
<DatePicker
  ids={
    Object {
      "day": "mock-day-id",
      "month": "mock-month-id",
      "year": "mock-year-id",
    }
  }
  onChange={[MockFunction]}
>
  <div
    className="c-datePicker"
    onBlur={[Function]}
    onFocus={[Function]}
  >
    <div
      className="row no-gutters pr-2"
    >
      <div
        className="col pl-2 pr-1"
      >
        <Input
          bredde="fullbredde"
          className="DatePickerDayInput"
          disabled={false}
          feil={false}
          id="mock-day-id"
          label="Dag"
          max="31"
          min="1"
          onChange={[Function]}
          onInput={[Function]}
          placeholder="DD"
          type="number"
          value=""
        >
          <div
            className="skjemaelement DatePickerDayInput"
          >
            <Label
              htmlFor="mock-day-id"
            >
              <label
                className="skjemaelement__label"
                htmlFor="mock-day-id"
              >
                Dag
              </label>
            </Label>
            <input
              aria-invalid={false}
              className="skjemaelement__input input--fullbredde"
              disabled={false}
              id="mock-day-id"
              max="31"
              min="1"
              onChange={[Function]}
              onInput={[Function]}
              placeholder="DD"
              type="number"
              value=""
            />
          </div>
        </Input>
      </div>
      <div
        className="col pl-1 pr-1"
      >
        <Input
          bredde="fullbredde"
          className="DatePickerMonthInput"
          disabled={false}
          feil={false}
          id="mock-month-id"
          label="Måned"
          max="12"
          min="1"
          onChange={[Function]}
          onInput={[Function]}
          placeholder="MM"
          type="number"
          value=""
        >
          <div
            className="skjemaelement DatePickerMonthInput"
          >
            <Label
              htmlFor="mock-month-id"
            >
              <label
                className="skjemaelement__label"
                htmlFor="mock-month-id"
              >
                Måned
              </label>
            </Label>
            <input
              aria-invalid={false}
              className="skjemaelement__input input--fullbredde"
              disabled={false}
              id="mock-month-id"
              max="12"
              min="1"
              onChange={[Function]}
              onInput={[Function]}
              placeholder="MM"
              type="number"
              value=""
            />
          </div>
        </Input>
      </div>
      <div
        className="col col pl-1 pr-0"
      >
        <Input
          bredde="fullbredde"
          className="DatePickerYearInput"
          disabled={false}
          feil={false}
          id="mock-year-id"
          label="År"
          min="1900"
          onChange={[Function]}
          onInput={[Function]}
          placeholder="ÅÅÅÅ"
          type="number"
          value=""
        >
          <div
            className="skjemaelement DatePickerYearInput"
          >
            <Label
              htmlFor="mock-year-id"
            >
              <label
                className="skjemaelement__label"
                htmlFor="mock-year-id"
              >
                År
              </label>
            </Label>
            <input
              aria-invalid={false}
              className="skjemaelement__input input--fullbredde"
              disabled={false}
              id="mock-year-id"
              min="1900"
              onChange={[Function]}
              onInput={[Function]}
              placeholder="ÅÅÅÅ"
              type="number"
              value=""
            />
          </div>
        </Input>
      </div>
    </div>
  </div>
</DatePicker>
`;
